File: backend/bin/src/controllers/appointment.controller.js
const { Appointment, Patient, Doctor, User, Payment, Prescription, Notification } = require('../models');
const { notifyUser } = require('../services/socket.service');
const { sendEmail } = require('../services/email.service');

const bookAppointment = async (req, res) => {
  try {
    const { doctor_id, date, time } = req.body;

    if (!doctor_id || !date || !time) {
      return res.status(400).json({ success: false, message: 'Doctor ID, date, and time are required.' });
    }

    // 1. Fetch Patient details
    const patient = await Patient.findOne({ where: { user_id: req.user.id } });
    if (!patient) {
      return res.status(404).json({ success: false, message: 'Patient profile not found.' });
    }

    // 2. Fetch Doctor details
    const doctor = await Doctor.findByPk(doctor_id, {
      include: [{ model: User, attributes: ['name', 'email'] }]
    });
    if (!doctor || doctor.verification_status !== 'approved') {
      return res.status(400).json({ success: false, message: 'Selected doctor is not available or verified.' });
    }

    // 3. Prevent duplicate scheduling conflicts
    const conflict = await Appointment.findOne({
      where: {
        doctor_id,
        date,
        time,
        status: ['pending', 'confirmed']
      }
    });

    if (conflict) {
      return res.status(409).json({ success: false, message: 'The doctor already has an appointment booked at this slot.' });
    }

    // 4. Create appointment
    const appointment = await Appointment.create({
      patient_id: patient.patient_id,
      doctor_id,
      date,
      time,
      status: 'pending'
    });

    // 5. Pre-create payment record (pending payment)
    await Payment.create({
      appointment_id: appointment.appointment_id,
      amount: doctor.consultation_fee,
      payment_status: 'pending'
    });

    // 6. Notify Doctor (in-app notification, socket event, and email)
    const alertTitle = 'New Appointment Request';
    const alertMsg = `Patient ${req.user.name} has requested an appointment on ${date} at ${time}.`;

    const notification = await Notification.create({
      user_id: doctor.user_id,
      title: alertTitle,
      message: alertMsg
    });

    // Emit live push event
    const io = req.app.get('io');
    notifyUser(io, doctor.user_id, notification);

    // Dispatch email
    await sendEmail({
      to: doctor.User.email,
      subject: alertTitle,
      text: alertMsg,
      html: `<p>${alertMsg}</p>`
    });

    res.status(201).json({
      success: true,
      message: 'Appointment booked successfully. Payment is pending.',
      data: appointment
    });
  } catch (error) {
    console.error('Book Appointment Error:', error);
    res.status(500).json({ success: false, message: 'Failed to book appointment.' });
  }
};

const listPatientAppointments = async (req, res) => {
  try {
    const patient = await Patient.findOne({ where: { user_id: req.user.id } });
    if (!patient) {
      return res.status(404).json({ success: false, message: 'Patient profile not found.' });
    }

    const appointments = await Appointment.findAll({
      where: { patient_id: patient.patient_id },
      include: [
        {
          model: Doctor,
          include: [{ model: User, attributes: ['name', 'email', 'id'] }]
        },
        { model: Payment },
        { model: Prescription }
      ],
      order: [['date', 'DESC'], ['time', 'DESC']]
    });

    res.status(200).json({ success: true, data: appointments });
  } catch (error) {
    console.error('List Patient Appointments Error:', error);
    res.status(500).json({ success: false, message: 'Failed to retrieve appointments.' });
  }
};

const listDoctorAppointments = async (req, res) => {
  try {
    const doctor = await Doctor.findOne({ where: { user_id: req.user.id } });
    if (!doctor) {
      return res.status(404).json({ success: false, message: 'Doctor profile not found.' });
    }

    const appointments = await Appointment.findAll({
      where: { doctor_id: doctor.doctor_id },
      include: [
        {
          model: Patient,
          include: [{ model: User, attributes: ['name', 'email', 'id'] }]
        },
        { model: Payment },
        { model: Prescription }
      ],
      order: [['date', 'DESC'], ['time', 'DESC']]
    });

    res.status(200).json({ success: true, data: appointments });
  } catch (error) {
    console.error('List Doctor Appointments Error:', error);
    res.status(500).json({ success: false, message: 'Failed to retrieve appointments.' });
  }
};

const updateAppointmentStatus = async (req, res) => {
  try {
    const { id } = req.params;
    const { status } = req.body;

    if (!['confirmed', 'completed', 'cancelled'].includes(status)) {
      return res.status(400).json({ success: false, message: 'Invalid status state.' });
    }

    const appointment = await Appointment.findByPk(id, {
      include: [
        { model: Patient, include: [{ model: User, attributes: ['name', 'email', 'user_id', 'id'] }] },
        { model: Doctor, include: [{ model: User, attributes: ['name', 'email', 'user_id', 'id'] }] }
      ]
    });

    if (!appointment) {
      return res.status(404).json({ success: false, message: 'Appointment not found.' });
    }

    // Role restrictions for updating state:
    // Patient can cancel. Doctor can confirm, complete, cancel.
    if (req.user.role === 'patient' && status !== 'cancelled') {
      return res.status(403).json({ success: false, message: 'Patients can only cancel appointments.' });
    }

    // Apply status update
    appointment.status = status;
    await appointment.save();

    // Trigger Notification for the other participant
    let targetUserId, targetEmail, alertTitle, alertMsg;
    if (req.user.role === 'patient') {
      // Patient updated status (cancelled it)
      targetUserId = appointment.Doctor.User.id;
      targetEmail = appointment.Doctor.User.email;
      alertTitle = 'Appointment Cancelled by Patient';
      alertMsg = `Appointment scheduled on ${appointment.date} was cancelled by Patient ${appointment.Patient.User.name}.`;
    } else {
      // Doctor updated status
      targetUserId = appointment.Patient.User.id;
      targetEmail = appointment.Patient.User.email;
      alertTitle = `Appointment status updated: ${status.toUpperCase()}`;
      alertMsg = `Doctor ${appointment.Doctor.User.name} has updated your appointment status to: ${status}.`;
    }

    const notification = await Notification.create({
      user_id: targetUserId,
      title: alertTitle,
      message: alertMsg
    });

    // Realtime push
    const io = req.app.get('io');
    notifyUser(io, targetUserId, notification);

    // Send email
    await sendEmail({
      to: targetEmail,
      subject: alertTitle,
      text: alertMsg,
      html: `<p>${alertMsg}</p>`
    });

    res.status(200).json({ success: true, message: 'Appointment status updated successfully.', data: appointment });
  } catch (error) {
    console.error('Update Appointment Status Error:', error);
    res.status(500).json({ success: false, message: 'Failed to update appointment.' });
  }
};

module.exports = {
  bookAppointment,
  listPatientAppointments,
  listDoctorAppointments,
  updateAppointmentStatus
};

File: backend/src/main/java/com/telehealth/pro/controller/AppointmentController.java
package com.telehealth.pro.controller;

import com.telehealth.pro.dto.ApiResponse;
import com.telehealth.pro.entity.*;
import com.telehealth.pro.repository.*;
import com.telehealth.pro.security.CustomUserDetails;
import com.telehealth.pro.service.EmailService;
import com.telehealth.pro.service.SocketService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/appointments")
public class AppointmentController {

    @Autowired
    private PatientRepository patientRepository;

    @Autowired
    private DoctorRepository doctorRepository;

    @Autowired
    private AppointmentRepository appointmentRepository;

    @Autowired
    private PaymentRepository paymentRepository;

    @Autowired
    private NotificationRepository notificationRepository;

    @Autowired
    private SocketService socketService;

    @Autowired
    private EmailService emailService;

    @Autowired
    private UserRepository userRepository;

    @PostMapping
    @PreAuthorize("hasRole('PATIENT')")
    @Transactional
    public ResponseEntity<?> bookAppointment(
            @AuthenticationPrincipal CustomUserDetails userDetails,
            @RequestBody Map<String, Object> body) {
        
        User currentUser = userDetails.getUser();
        
        if (body.get("doctor_id") == null || body.get("date") == null || body.get("time") == null) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(ApiResponse.error("Doctor ID, date, and time are required."));
        }

        Long doctorId = ((Number) body.get("doctor_id")).longValue();
        String dateStr = (String) body.get("date");
        String timeStr = (String) body.get("time");

        Optional<Patient> patientOpt = patientRepository.findByUser_Id(currentUser.getId());
        if (patientOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse.error("Patient profile not found."));
        }
        Patient patient = patientOpt.get();

        Optional<Doctor> doctorOpt = doctorRepository.findById(doctorId);
        if (doctorOpt.isEmpty() || !"approved".equals(doctorOpt.get().getVerificationStatus())) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(ApiResponse.error("Selected doctor is not available or verified."));
        }
        Doctor doctor = doctorOpt.get();

        LocalDate date = LocalDate.parse(dateStr);
        LocalTime time = LocalTime.parse(timeStr);

        // Check scheduling conflicts
        Optional<Appointment> conflict = appointmentRepository.findConflict(doctorId, date, time, Arrays.asList("pending", "confirmed"));
        if (conflict.isPresent()) {
            return ResponseEntity.status(HttpStatus.CONFLICT)
                    .body(ApiResponse.error("The doctor already has an appointment booked at this slot."));
        }

        Appointment appointment = Appointment.builder()
                .patient(patient)
                .doctor(doctor)
                .date(date)
                .time(time)
                .status("pending")
                .build();

        appointment = appointmentRepository.save(appointment);

        // Pre-create payment record
        Payment payment = Payment.builder()
                .appointment(appointment)
                .amount(doctor.getConsultationFee())
                .paymentStatus("pending")
                .build();
        paymentRepository.save(payment);

        // Notification for Doctor
        String alertTitle = "New Appointment Request";
        String alertMsg = "Patient " + currentUser.getName() + " has requested an appointment on " + dateStr + " at " + timeStr + ".";

        Notification notification = Notification.builder()
                .user(doctor.getUser())
                .title(alertTitle)
                .message(alertMsg)
                .build();
        notification = notificationRepository.save(notification);

        // Realtime WebSockets push & email dispatch
        socketService.notifyUser(doctor.getUser().getId(), notification);
        emailService.sendEmail(doctor.getUser().getEmail(), alertTitle, alertMsg, "<p>" + alertMsg + "</p>");

        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success("Appointment booked successfully. Payment is pending.", appointment));
    }

    @GetMapping("/patient")
    @PreAuthorize("hasRole('PATIENT')")
    public ResponseEntity<?> listPatientAppointments(@AuthenticationPrincipal CustomUserDetails userDetails) {
        User currentUser = userDetails.getUser();
        Optional<Patient> patientOpt = patientRepository.findByUser_Id(currentUser.getId());
        if (patientOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse.error("Patient profile not found."));
        }
        
        List<Appointment> appointments = appointmentRepository
                .findAllByPatientPatientIdOrderByDateDescTimeDesc(patientOpt.get().getPatientId());
        return ResponseEntity.ok(ApiResponse.success(appointments));
    }

    @GetMapping("/doctor")
    @PreAuthorize("hasRole('DOCTOR')")
    public ResponseEntity<?> listDoctorAppointments(@AuthenticationPrincipal CustomUserDetails userDetails) {
        User currentUser = userDetails.getUser();
        Optional<Doctor> doctorOpt = doctorRepository.findByUser_Id(currentUser.getId());
        if (doctorOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse.error("Doctor profile not found."));
        }

        List<Appointment> appointments = appointmentRepository
                .findAllByDoctorDoctorIdOrderByDateDescTimeDesc(doctorOpt.get().getDoctorId());
        return ResponseEntity.ok(ApiResponse.success(appointments));
    }

    @PutMapping("/{id}/status")
    @PreAuthorize("hasAnyRole('PATIENT', 'DOCTOR')")
    @Transactional
    public ResponseEntity<?> updateAppointmentStatus(
            @AuthenticationPrincipal CustomUserDetails userDetails,
            @PathVariable Long id,
            @RequestBody Map<String, String> body) {
        
        User currentUser = userDetails.getUser();
        String status = body.get("status");

        if (!Arrays.asList("confirmed", "completed", "cancelled").contains(status)) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(ApiResponse.error("Invalid status state."));
        }

        Optional<Appointment> appointmentOpt = appointmentRepository.findById(id);
        if (appointmentOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse.error("Appointment not found."));
        }

        Appointment appointment = appointmentOpt.get();

        if ("patient".equals(currentUser.getRole()) && !"cancelled".equals(status)) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN)
                    .body(ApiResponse.error("Patients can only cancel appointments."));
        }

        appointment.setStatus(status);
        appointment = appointmentRepository.save(appointment);

        // Notify other user
        Long targetUserId;
        String targetEmail;
        String alertTitle;
        String alertMsg;

        if ("patient".equals(currentUser.getRole())) {
            targetUserId = appointment.getDoctor().getUser().getId();
            targetEmail = appointment.getDoctor().getUser().getEmail();
            alertTitle = "Appointment Cancelled by Patient";
            alertMsg = "Appointment scheduled on " + appointment.getDate() + " was cancelled by Patient " + appointment.getPatient().getUser().getName() + ".";
        } else {
            targetUserId = appointment.getPatient().getUser().getId();
            targetEmail = appointment.getPatient().getUser().getEmail();
            alertTitle = "Appointment status updated: " + status.toUpperCase();
            alertMsg = "Doctor " + appointment.getDoctor().getUser().getName() + " has updated your appointment status to: " + status + ".";
        }

        Notification notification = Notification.builder()
                .user(userRepository.getReferenceById(targetUserId))
                .title(alertTitle)
                .message(alertMsg)
                .build();
        notification = notificationRepository.save(notification);

        // Realtime WebSockets push & email dispatch
        socketService.notifyUser(targetUserId, notification);
        emailService.sendEmail(targetEmail, alertTitle, alertMsg, "<p>" + alertMsg + "</p>");

        return ResponseEntity.ok(ApiResponse.success("Appointment status updated successfully.", appointment));
    }
}
